### PR TITLE
[LoadTest] Revamp load test suite

### PR DESCRIPTION
**Pull request by red-0ne**
## Summary

This pull request includes several updates to the load-testing configuration and test files, as well as improvements to the relay stress test suite. The most important changes include updating funding account addresses, modifying gateway configurations, and refactoring various test functions for better accuracy.

## Type of change

Select one or more from the following:

- [ ] New feature, functionality or library
- [ ] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [ ] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

**Comment by Olshansk** [comment]
@red-0ne Lmk if you need me to take a look at this as well.

File: load-testing/tests/relays_stress_helpers_test.go [hunk 1]
```diff
@@ -75,28 +81,46 @@ type actorLoadTestIncrementPlan struct {
...
+	channel.ForEach(
 		s.ctx,
-		deps,
-		newTxEventSubscriptionQuery,
-		tx.UnmarshalTxResult,
-		eventsReplayClientBufferSize,
-	)
-	require.NoError(s, err)
+		s.blockClient.CommittedBlocksSequence(s.ctx),
+		func(ctx context.Context, block client.Block) {
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:96]: NIT: Can you assign this to a locally scope function so the `.ForEach` is easier to read?

File: load-testing/tests/relays_stress_helpers_test.go [hunk 2]
```diff
@@ -631,7 +651,8 @@ func (s *relaysSuite) getAppFundingAmount(currentBlockHeight int64) sdk.Coin {
 	// based on the number of relays it needs to send. Theoretically, `+1` should
 	// be enough, but probabilistic and time based mechanisms make it hard
 	// to predict exactly.
-	appFundingAmount := s.relayRatePerApp * s.relayCoinAmountCost * currentTestDuration * blockDuration * 2
+	amountToSpend := s.relayRatePerApp * s.relayCoinAmountCost * currentTestDuration * blockDuration
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:654]:
```suggestion
	relayCostDuringTestPOKT := s.relayRatePerApp * s.relayCoinAmountCost * currentTestDuration * blockDuration
```

File: load-testing/tests/relays_stress_helpers_test.go [hunk 3]
```diff
@@ -631,7 +651,8 @@ func (s *relaysSuite) getAppFundingAmount(currentBlockHeight int64) sdk.Coin {
...
+	appFundingAmount := math.Max(amountToSpend, s.appParams.MinStake.Amount.Int64()*2)
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:655]: Move the comment starting with `Multiply by 2` above so it's above this line instead.

File: load-testing/tests/relays_stress_helpers_test.go [hunk 4]
```diff
@@ -1034,150 +1031,189 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
 	// Send the relay request within a goroutine to avoid blocking the test batches
 	// when suppliers or gateways are unresponsive.
-	go func(gwURL, payload string) {
-		_, err = http.DefaultClient.Post(
-			gwURL,
-			"application/json",
-			strings.NewReader(payload),
-		)
+	go func(gwURL, appAddr, payload string) {
+		gwURL = gwURL + "/"
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1038]: Can we just call it `gatewayURL` without the shortening?

File: load-testing/tests/relays_stress_helpers_test.go [hunk 5]
```diff
@@ -1034,150 +1031,189 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
+		req, err := http.NewRequest("POST", gwURL, strings.NewReader(payload))
+
+		req.Header.Add("X-App-Address", appAddr)
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1041]: 1. TODO_TECHDEBT to remove `x-`
2. #PUC why we need this

File: load-testing/tests/relays_stress_helpers_test.go [hunk 6]
```diff
@@ -1034,150 +1031,189 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
+		res, err := http.DefaultClient.Do(req)
 		require.NoError(s, err)
-	}(gatewayUrl.String(), relayPayload)
+
+		if res.StatusCode != http.StatusOK {
+			s.failedRelays.Add(1)
+		}
+	}(s.gatewayUrls[gateway.address], application.address, relayPayload)
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1048]: What if you do something like:

```go
inlineFunc := func(...) {..}
gatewayURL := s.gatewayUrls[...]
go inlineFunc(gatewayURL, ...)
```

You'll need better names, but it'll be easier to maintain.,

**Review by Olshansk — CHANGES_REQUESTED** [review]
@red-0ne Partial review.

This is great work but `load-testing/tests/relays_stress_helpers_test.go` is really hard to review right now with lots of big inline functions.

Do you mind doing a self passthrough and finding ways to make it a bit more readable & maintainable for others?

**Comment by red-0ne** [comment]
> @red-0ne Partial review.
> 
> This is great work but `load-testing/tests/relays_stress_helpers_test.go` is really hard to review right now with lots of big inline functions.
> 
> Do you mind doing a self passthrough and finding ways to make it a bit more readable & maintainable for others?

It is indeed very hard to follow, we have a TODO_TECHDEBT where we'll have to address this particular issue:

image

Will create an issue and prioritize accordingly.

File: load-testing/tests/relays_stress.feature [hunk 7]
```diff
@@ -14,4 +14,12 @@ Feature: Loading gateway server with relays
       | gateway     | 1                | 10             | 3          |
       | supplier    | 1                | 10             | 3          |
     When a load of concurrent relay requests are sent from the applications
-    Then the correct pairs count of claim and proof messages should be committed on-chain
\ No newline at end of file
+    Then the number of failed relay requests is "0"
+    # TODO_FOLLOWUP(@red-0ne): Implement the following steps
+    # Then "0" over servicing events are observed
+    # And "0" slashing events are observed
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply load-testing/tests/relays_stress.feature:20]: The load test should not exercise over servicing or slashing but rather always assert that none of them happen given a well functioning RelayMiner.

Additionally, in order to "force" the RelayMiner to have this behavior; we'll have to implement the functionality or set up the right environment to do so which I think would be in the scope of the E2E tests.

File: load-testing/tests/relays_stress_helpers_test.go [hunk 8]
```diff
@@ -75,29 +81,49 @@ type actorLoadTestIncrementPlan struct {
...
 }
 
-// setupTxEventListeners sets up the transaction event listeners to observe the
-// transactions committed on-chain.
-func (s *relaysSuite) setupTxEventListeners() {
-	eventsQueryClient := testeventsquery.NewLocalnetClient(s.TestingT.(*testing.T))
+// setupEventListeners sets up the event listeners for the relays suite.
+// It listens to both tx and block events to keep track of the events that are happening
+// on the chain.
+func (s *relaysSuite) setupEventListeners(pocketNode string) {
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:87]: s/pocketNode/rpcNode?

File: load-testing/tests/relays_stress_helpers_test.go [hunk 9]
```diff
@@ -75,29 +81,49 @@ type actorLoadTestIncrementPlan struct {
...
+	// Set up the blockClient that will be notifying the suite about the committed blocks.
+	eventsObs, eventsObsCh := channel.NewObservable[[]types.Event]()
+	s.eventsObs = eventsObs
+
+	extractBlockEvents := func(ctx context.Context, block client.Block) {
+		// Query the block results endpoint for each observed block to get the tx and block events.
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:93]:
```suggestion
		// Query the block results endpoint for each observed block to get the tx and block events.
		// Ref: https://docs.cometbft.com/main/rpc/#/Info/block_results
```

File: load-testing/tests/relays_stress_helpers_test.go [hunk 10]
```diff
@@ -75,29 +81,49 @@ type actorLoadTestIncrementPlan struct {
...
+
+		var rpcResponse rpctypes.RPCResponse
+		err = json.Unmarshal(blockResultsRespBz, &rpcResponse)
+		require.NoError(s, err)
+
+		var blockResults cmtcoretypes.ResultBlockResults
+		err = json.Unmarshal(rpcResponse.Result, &blockResults)
+		require.NoError(s, err)
+
+		events := make([]types.Event, 0, len(blockResults.TxsResults)+len(blockResults.FinalizeBlockEvents))
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:111]:
```suggestion
		numEvents := len(blockResults.TxsResults)+len(blockResults.FinalizeBlockEvents)
		events := make([]types.Event, 0, numEvents)
```

File: load-testing/tests/relays_stress_helpers_test.go [hunk 11]
```diff
@@ -75,29 +81,49 @@ type actorLoadTestIncrementPlan struct {
...
+		for _, txResult := range blockResults.TxsResults {
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:112]:
```suggestion
		// Flatten all tx result events and block event results into one slice
		for _, txResult := range blockResults.TxsResults {
```

File: load-testing/tests/relays_stress_helpers_test.go [hunk 12]
```diff
@@ -627,11 +649,12 @@ func (s *relaysSuite) createApplicationAccount(
 // cost, and the block duration.
 func (s *relaysSuite) getAppFundingAmount(currentBlockHeight int64) sdk.Coin {
 	currentTestDuration := s.testStartHeight + s.relayLoadDurationBlocks - currentBlockHeight
+	relayCostDuringTestUPOKT := s.relayRatePerApp * s.relayCoinAmountCost * currentTestDuration * blockDuration
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:652]:
```suggestion
	// Compute the cost of all relays throughout the test duration
	totalRelayCostDuringTestUPOKT := s.relayRatePerApp * s.relayCoinAmountCost * currentTestDuration * blockDuration
```

File: load-testing/tests/relays_stress_helpers_test.go [hunk 13]
```diff
@@ -943,7 +969,7 @@ func (s *relaysSuite) sendPendingMsgsTx(actor *accountInfo) {
 	err := txBuilder.SetMsgs(actor.pendingMsgs...)
 	require.NoError(s, err)
 
-	txBuilder.SetTimeoutHeight(uint64(s.latestBlock.Height() + 1))
+	txBuilder.SetTimeoutHeight(uint64(s.latestBlock.Height() + 2))
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:972]: #PUC why this is +2

File: load-testing/tests/relays_stress_helpers_test.go [hunk 14]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
 	// Send the relay request within a goroutine to avoid blocking the test batches
 	// when suppliers or gateways are unresponsive.
-	go func(gwURL, payload string) {
-		_, err = http.DefaultClient.Post(
-			gwURL,
-			"application/json",
-			strings.NewReader(payload),
-		)
+	sendRelayRequest := func(gatewayURL, appAddr, payload string) {
+		gatewayURL = gatewayURL + "/"
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1040]: 1. #PUC why we need to add "/"
2. Should we not check if there's a trailing / first? Maybe the std string library has something for it

File: load-testing/tests/relays_stress_helpers_test.go [hunk 15]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
+		// This is needed by the PATH Gateway's trusted mode to identify the application
+		// that is sending the relay request.
+		req.Header.Add("X-App-Address", appAddr)
+		req.Header.Add("target-service-id", appAddr)
+		res, err := http.DefaultClient.Do(req)
 		require.NoError(s, err)
-	}(gatewayUrl.String(), relayPayload)
+
+		if res.StatusCode != http.StatusOK {
+			s.failedRelays.Add(1)
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1053]: 1. Do we count successful relays anywhere?  If not, could be good for telemetry, printing & debugging later.
2. What do you think of doing a print statement every N (e.g. 100) relays like:

e.g. Sent: X relays; Success: Y; Failed: Z

File: load-testing/tests/relays_stress_helpers_test.go [hunk 16]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
-				}
-
-				attrs := event.Attributes
-				// Check if the actor is the recipient of the transfer event.
-				if actorFunded = hasEventAttr(attrs, "recipient", actor.address); actorFunded {
-					break
-				}
+func (s *relaysSuite) ensureFundedActors(ctx context.Context, actors []*accountInfo) {
+	if len(actors) == 0 {
+		return
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1067]: log a warning?

File: load-testing/tests/relays_stress_helpers_test.go [hunk 17]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
+	}
+
+	fundedActors := make(map[string]struct{})
+	actorsAddrs := make([]string, len(actors))
+	for i, actor := range actors {
+		actorsAddrs[i] = actor.address
+	}
+
+	deadline := time.Now().Add(time.Second * time.Duration(blockDuration+1))
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1076]: NITs:

1. Should we rename `blockDuration` to `blockDurationSec`?
2. #PUC why +1

File: load-testing/tests/relays_stress_helpers_test.go [hunk 18]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
+	ctx, cancel := context.WithDeadline(ctx, deadline)
+	channel.ForEach(ctx, s.eventsObs, func(ctx context.Context, events []types.Event) {
+		for _, event := range events {
+			// Skip non-relevant events.
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1080]: #PUC why `transfer` is the only relevant event

File: load-testing/tests/relays_stress_helpers_test.go [hunk 19]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
 		}
+	})
+
+	<-ctx.Done()
+	if !allActorsFunded(actors, fundedActors) {
+		s.logAndAbortTest("actors not funded")
 	}
 }
 
+// allActorsFunded checks if all the expected actors are funded.
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1110]:
```suggestion
// allActorsFunded checks if all the expected actors are funded.
// An error is returned if at least one of the expected actors was not funded.
```

File: load-testing/tests/relays_stress_helpers_test.go [hunk 20]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
-			s.logAndAbortTest(txResults, "actor not funded")
-			return
+		if allActorsFunded(actors, fundedActors) {
+			cancel()
 		}
+	})
+
+	<-ctx.Done()
+	if !allActorsFunded(actors, fundedActors) {
+		s.logAndAbortTest("actors not funded")
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1106]:
```suggestion
		s.logAndAbortTest("at least one actor was not funded successfully")
```

File: load-testing/tests/relays_stress_helpers_test.go [hunk 21]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
+	if len(actors) == 0 {
+		return
+	}
 
-			// If the actor is staked, no need to check the other transactions.
-			if actorStaked {
-				break
+	stakedActors := make(map[string]struct{})
+
+	deadline := time.Now().Add(time.Second * time.Duration(blockDuration+1))
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1133]: Ditto

File: load-testing/tests/relays_stress_helpers_test.go [hunk 22]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
+				stakedActors[e.Application.GetAddress()] = struct{}{}
 			}
 		}
 
-		// If no message event is found for the actor, log the transaction results
-		// and cancel the test.
-		if !actorStaked {
-			s.logAndAbortTest(txResults, fmt.Sprintf("actor not staked: %s", actor.address))
-			return
+		if allActorsStaked(actors, stakedActors) {
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1148]:
```suggestion
		// We can cancel exit this scope once all expected actors were successfully staked
		// before the deadline was reached.
		if allActorsStaked(actors, stakedActors) {
```

File: load-testing/tests/relays_stress_helpers_test.go [hunk 23]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
+	})
+
+	<-ctx.Done()
+	if !allActorsStaked(actors, stakedActors) {
+		s.logAndAbortTest("actors not staked")
+		return
 	}
 }
 
+// allActorsStaked checks if all the expected actors are staked.
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1160]:
```suggestion
// allActorsStaked checks if all the expected actors are staked.
// An error is returned if at least one of the expected actors was not staked.
```

File: load-testing/tests/relays_stress_helpers_test.go [hunk 24]
```diff
@@ -1034,148 +1033,195 @@ func (s *relaysSuite) sendRelay(iteration uint64, relayPayload string) (appAddre
...
-			s.logAndAbortTest(txResults, fmt.Sprintf("actor not staked: %s", actor.address))
-			return
+		if allActorsStaked(actors, stakedActors) {
+			cancel()
 		}
+	})
+
+	<-ctx.Done()
+	if !allActorsStaked(actors, stakedActors) {
+		s.logAndAbortTest("actors not staked")
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1155]:
```suggestion
		s.logAndAbortTest("at least one actor was not staked successfully")
```

File: load-testing/tests/relays_stress_test.go [hunk 25]
```diff
@@ -336,33 +314,26 @@ func (s *relaysSuite) LocalnetIsRunning() {
...
-
 	// Setup the txContext that will be used to send transactions to the network.
 	s.txContext = testtx.NewLocalnetContext(s.TestingT.(*testing.T))
 
 	// Get the relay cost from the tokenomics module.
 	s.relayCoinAmountCost = s.getRelayCost()
 
-	// Setup the tx listener for on-chain events to check and assert on transactions results.
-	s.setupTxEventListeners()
+	// Setup the event listener for on-chain events to check and assert on transactions
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_test.go:323]: s/on-chain/onchain

https://x.com/olshansky/status/1869115777800180217

Ditto everywhere else (e.g. line 330 below) but I won't leave any more comments.

Please do a find-all "on-chain"

File: load-testing/tests/relays_stress_test.go [hunk 26]
```diff
@@ -430,11 +390,11 @@ func (s *relaysSuite) MoreActorsAreStakedAsFollows(table gocuke.DataTable) {
 	gateways := fundedGateways[:s.gatewayInitialCount]
 	applications := fundedApplications[:s.appInitialCount]
 
+	stakedActors := append(suppliers, gateways...)
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_test.go:393]: Trying to be consistent with "funded" above.

```suggestion
	stakedActors := append(stakedSuppliers, stakedGateways...)
```

File: load-testing/tests/relays_stress_test.go [hunk 27]
```diff
@@ -450,8 +410,7 @@ func (s *relaysSuite) MoreActorsAreStakedAsFollows(table gocuke.DataTable) {
 
 	// Delegate the initial applications to the initial gateways
 	s.sendDelegateInitialAppsTxs(applications, gateways)
-	txResults = s.waitForTxsToBeCommitted()
-	s.ensureDelegatedApps(txResults, applications, gateways)
+	s.ensureDelegatedApps(s.ctx, applications, gateways)
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_test.go:413]: Ditto. consider adding a prefix

File: load-testing/tests/relays_stress_helpers_test.go [hunk 28]
```diff
@@ -1479,3 +1570,36 @@ func (s *relaysSuite) populateWithKnownGateways() (gateways []*accountInfo) {
 
 	return gateways
 }
+
+// WaitAll waits for all the provided functions to complete.
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1574]: Nicely done!

File: load-testing/tests/relays_stress_helpers_test.go [hunk 29]
```diff
@@ -1479,3 +1570,36 @@ func (s *relaysSuite) populateWithKnownGateways() (gateways []*accountInfo) {
...
+			f()
+			wg.Done()
+		}(f)
+	}
+
+	wg.Wait()
+}
+
+// abciEventsToTypedEvents converts the abci events to typed events.
+func abciEventsToTypedEvents(ctx context.Context, abciEventObs observable.Observable[[]types.Event]) observable.Observable[[]proto.Message] {
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1591]: OPTIONAL: What do you think of moving this into `poktroll/testutil/events/filter.go`?

File: load-testing/tests/relays_stress_helpers_test.go [hunk 30]
```diff
@@ -1479,3 +1570,36 @@ func (s *relaysSuite) populateWithKnownGateways() (gateways []*accountInfo) {
 
 	return gateways
 }
+
+// WaitAll waits for all the provided functions to complete.
+// It is used to wait for multiple goroutines to complete before proceeding.
+func (s *relaysSuite) WaitAll(waitFuncs ...func()) {
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1576]: I noticed this doesn't use `s` at all. 

Can we move this into TestUtil?

File: load-testing/tests/relays_stress_single_supplier.feature [hunk 31]
```diff
@@ -11,7 +11,15 @@ Feature: Loading gateway server with relays
...
       | supplier    | 1                | 10             | 1          |
     When a load of concurrent relay requests are sent from the applications
-    Then the correct pairs count of claim and proof messages should be committed on-chain
\ No newline at end of file
+    Then the number of failed relay requests is "0"
+    # TODO_FOLLOWUP(@red-0ne): Implement the following steps
+    # Then "0" over servicing events are observed
+    # And "0" slashing events are observed
+    # And "0" expired claim events are observed
+    # And there is as many reimbursement requests as the number of settled claims
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_single_supplier.feature:22]:
```suggestion
    # And there are as many reimbursement requests as the number of settled claims
```

File: load-testing/tests/relays_stress_single_supplier.feature [hunk 32]
```diff
@@ -11,7 +11,15 @@ Feature: Loading gateway server with relays
...
+    # And the number of claims submitted and claims settled is the same
+    # And the number of proofs submitted and proofs required is the same
+    # And the actors onchain balances are as expected
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_single_supplier.feature:25]: Note to our selves: revisit some more interesting test cases here.

File: load-testing/tests/relays_stress_test.go [hunk 33]
```diff
@@ -213,17 +193,15 @@ type relaysSuite struct {
...
-
 	// isEphemeralChain is a flag that indicates whether the test is expected to be
 	// run on ephemeral chain setups like localnet or long living ones (i.e. Test/DevNet).
 	isEphemeralChain bool
+
+	// eventsObs is the observable that maps committed blocks to on-chain events.
+	eventsObs observable.Observable[[]types.Event]
+
+	// failedRelays is the number of relay requests that returned non-200 status code.
+	failedRelays atomic.Uint64
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_test.go:204]: Per my other comment, considering counting successes as well.

File: load-testing/tests/relays_stress_helpers_test.go [hunk 34]
```diff
@@ -1379,19 +1411,83 @@ func (s *relaysSuite) querySharedParams(queryNodeRPCURL string) {
...
+	deps := depinject.Supply(s.txContext.GetClientCtx())
+
+	blockQueryClient, err := sdkclient.NewClientFromNode(queryNodeRPCURL)
+	require.NoError(s, err)
+	deps = depinject.Configs(deps, depinject.Supply(blockQueryClient))
+
+	appQueryclient, err := query.NewApplicationQuerier(deps)
+	require.NoError(s, err)
+
+	appParams, err := appQueryclient.GetParams(s.ctx)
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1428]: Why do all the other params need casting but this one doesn't?

File: load-testing/tests/relays_stress_helpers_test.go [hunk 35]
```diff
@@ -1379,19 +1411,83 @@ func (s *relaysSuite) querySharedParams(queryNodeRPCURL string) {
...
+func (s *relaysSuite) queryTokenomicsParams(queryNodeRPCURL string) {
+	s.Helper()
+
+	deps := depinject.Supply(s.txContext.GetClientCtx())
+
+	blockQueryClient, err := sdkclient.NewClientFromNode(queryNodeRPCURL)
+	require.NoError(s, err)
+	deps = depinject.Configs(deps, depinject.Supply(blockQueryClient))
+
+	var clientConn *grpc.ClientConn
```
Comment (Olshansk) [inline load-testing/tests/relays_stress_helpers_test.go:1468]: Why is querying tokenomic params differen than proof and app params?

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: load-testing/tests/relays_stress_helpers_test.go (hunk 30 above)
Comment (red-0ne) [reply load-testing/tests/relays_stress_helpers_test.go:1576]: Good catch!

File: load-testing/tests/relays_stress_helpers_test.go (hunk 34 above)
Comment (red-0ne) [reply load-testing/tests/relays_stress_helpers_test.go:1428]: The `proofParams` need casting because its querier returns an interface [1]; which was added to avoid dependency cycles.

[1] https://github.com/pokt-network/poktroll/blob/main/pkg/client/interface.go#L335

File: load-testing/tests/relays_stress_helpers_test.go (hunk 35 above)
Comment (red-0ne) [reply load-testing/tests/relays_stress_helpers_test.go:1468]: The reason we're using the auto-generated query client is that we don't have the need to implement a tokenomics client querier for now since offchain packages don't need it (yet).

Added a `TODO_TECHDEBT` to use the tokenomics querier once implemented.

File: load-testing/tests/relays_stress_test.go (hunk 25 above)
Comment (red-0ne) [reply load-testing/tests/relays_stress_test.go:325]: Replaced all the `on-chain/off-chain` occurrences with `onchain/offchain` but seemed too much changes for this already big PR.

Addressed them #1020

**Review by Olshansk — APPROVED** [review]
Just a few minor NITs but lets get it in afterwards!

File: load-testing/tests/relays_stress_test.go [hunk 36]
```diff
@@ -333,36 +313,29 @@ func (s *relaysSuite) LocalnetIsRunning() {
...
+	// Setup the event listener for on-chain events to check and assert on transactions
+	// and finalized blocks results.
+	s.setupEventListeners(loadTestParams.PRCNode)
 
 	// Initialize the funding account.
 	s.initFundingAccount(loadTestParams.FundingAccountAddress)
 
-	// Initialize the on-chain claims and proofs counter.
-	s.countClaimAndProofs()
+	// Initialize the on-chain settlement events listener.
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply load-testing/tests/relays_stress_test.go:332]: Fixed on PR #1020

File: testutil/delays/waitall.go [hunk 37]
```diff
@@ -0,0 +1,19 @@
...
+
+	for _, f := range waitFuncs {
+		go func(f func()) {
+			f()
+			wg.Done()
+		}(f)
+	}
+
+	wg.Wait()
+}
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply testutil/delays/waitall.go:19]: Neat!
